daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DaffodilUnparseContentHandler.scala

```diff
@@ -0,0 +1,309 @@
+ * contain no new information, until the unparse is completed.
+ * -  this process continues until the EndDocument method is called. Once that infosetEvent is
+ * sent to the inputter, it signals the end of events coming from the contentHandler. This
+ * ends the unparseProcess and returns the event with the unparseResult and/or any error
+ * information
+ *
+ * @param dp dataprocessor object that will be used to call the parse
+ * @param output outputChannel of choice where the unparsed data is stored
+ */
+class DaffodilUnparseContentHandler(
+  dp: DFDL.DataProcessor,
+  output: DFDL.Output)
+  extends DFDL.DaffodilUnparseContentHandler {
+  private lazy val inputter = new SAXInfosetInputter(this, dp, output)
+  private var unparseResult: DFDL.UnparseResult = _
+  private lazy val infosetEvent: DFDL.SaxInfosetEvent = new DFDL.SaxInfosetEvent
+  private lazy val characterData = new StringBuilder
+  private var prefixMapping: NamespaceBinding = _
+  private lazy val prefixMappingTrackingStack = new MStackOf[NamespaceBinding]
+  private var contentHandlerPrefixMappingUsed = false
+
+  /**
+   * returns null in the case of an DaffodilUnhandledSAXException
+   */
+  def getUnparseResult: DFDL.UnparseResult = unparseResult
+
+  def enableInputterResolutionOfRelativeInfosetBlobURIs(): Unit = inputter.enableResolutionOfRelativeInfosetBlobURIs()
+
+  override def setDocumentLocator(locator: Locator): Unit = {
+    // do nothing
+  }
+
+  override def startDocument(): Unit = {
+    infosetEvent.eventType = One(StartDocument)
+    sendToInputter()
+  }
+
+  override def endDocument(): Unit = {
+    infosetEvent.eventType = One(EndDocument)
+    sendToInputter()
+  }
+
+  override def startPrefixMapping(prefix: String, uri: String): Unit = {
+    contentHandlerPrefixMappingUsed = true
+    val pre = if (prefix == "") null else prefix
+    prefixMapping = NamespaceBinding(pre, uri, prefixMapping)
+  }
+
+  /**
+   * XMLReader does not guarantee the order of the prefixes called for this function, but it does
+   * guarantee that this method is called after its corresponding endElement, which means we can
+   * can just take off the top mappings, because the element that might have cared about the order
+   * is already done using the prefixMappings
+   */
+  override def endPrefixMapping(prefix: String): Unit = {
+    prefixMapping = if (prefixMapping == null) prefixMapping else prefixMapping.parent
+  }
+
+  /**
+   * Uses Attributes, which is passed in to the startElement callback, to extract prefix mappings and
+   * populate the global prefixMapping
+   */
+  def mapPrefixMappingFromAttributesImpl(atts:Attributes): Unit = {
+    var i = 0
+    while (i < atts.getLength) {
+      val qName = atts.getQName(i)
+      if (qName.startsWith("xmlns")) {
+        val uri =  atts.getValue(i)
+        val prefix = if(qName.contains(":")) {
+          val pref = qName.split(":").last
+          pref
+        } else {
```

Comment:
I think there's potential for very subtle, though rare, bugs. Which are some of the worst. For example, what if we have an attribute thats ``xmlnsblah:foo=whatever``. That's not an xmlns attribute, but this logic would think it is. I think maybe instead we need something more like
```scala
if (qname == "xmlns") {
  prefixMapping = NamespaceBinding(null, uri, prefixMapping) // no prefix
} else if (qname.startsWith("xmlns:")) {
  prefixMapping = NamespaceBinding(qname.substring(6), uri, prefixMapping)
} else {
  // no-op, not a namespace mapping attribute
}
```
This also avoids ``contains`` and ``split``, which I think are relatively expensive as far as string functions go--I think they both treat the string parameter as a regex behind the scenes.